runtime/jailer.go

```diff
@@ -0,0 +1,304 @@
+				fileName := filepath.Base(drivePath)
+
+				if err := os.Link(drivePath, filepath.Join(contentsPath, fileName)); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				newDrivePath := filepath.Join(strings.TrimPrefix(contentsPath, rootPath), fileName)
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(newDrivePath)
+			}
+
+			if err := chownr(
+				rootPath,
+				firecracker.IntValue(cfg.JailerConfig.UID),
+				firecracker.IntValue(cfg.JailerConfig.GID),
+			); err != nil {
+				return errors.Wrapf(err, "failed to chown -r %v", rootPath)
+			}
+
+			logger.Info("Successfully ran jailer handler")
+			return nil
+		},
+	}
+}
+
+// BuildLinkFifoHandler will return a new firecracker.Handler with the function
+// that will allow linking of the fifos making them visible to Firecracker.
+func (j jailer) BuildLinkFifoHandler() firecracker.Handler {
+	return firecracker.Handler{
+		Name: jailerFifoHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			contentsPath := j.ContentsPath()
+			fifoFileName := filepath.Base(m.Cfg.LogFifo)
+			newFifoPath := filepath.Join(contentsPath, fifoFileName)
+			if err := os.Link(m.Cfg.LogFifo, newFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.LogFifo = newFifoPath
+
+			metricFifoFileName := filepath.Base(m.Cfg.MetricsFifo)
+			newMetricFifoPath := filepath.Join(contentsPath, metricFifoFileName)
+			if err := os.Link(m.Cfg.MetricsFifo, newMetricFifoPath); err != nil {
+				return err
+			}
+			m.Cfg.MetricsFifo = newMetricFifoPath
+
+			return nil
+		},
+	}
+}
+
+func copyFile(src, dst string) error {
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", src)
+	}
+	defer srcFile.Close()
+
+	if info, err := os.Stat(dst); err == nil {
+		// no error indicates the path already exists, we check to see if it is a
+		// directory we will concatenate the file name provided in source.
+		if info.IsDir() {
+			dst = filepath.Join(dst, filepath.Base(src))
+		}
+	} else if os.IsNotExist(err) {
+		// this occurs when a full path is provided including the file name
+	} else if err != nil {
+		return err
+	}
+
+	dstFile, err := os.OpenFile(dst, os.O_CREATE|os.O_WRONLY, os.ModePerm)
+	if err != nil {
+		return errors.Wrapf(err, "failed to open %v", dstFile)
+	}
+	defer dstFile.Close()
+
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to copy to destination")
+	}
+	return nil
+}
+
+func chownr(path string, uid, gid int) error {
+	return filepath.Walk(path, func(name string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		return os.Chown(name, uid, gid)
+	})
+}
+
+func (j jailer) jailerCommand(containerName string) *exec.Cmd {
+	cmd := exec.Command("runc", "run", containerName)
+	cmd.Dir = j.JailPath()
+	return cmd
+}
+
+// overwriteConfig will set the proper default values if a field had not been set.
```

Comment:
This will overwrite the configuration values like `uid` and `gid`. In addition when we support numa nodes, that will also be injected into the config. So I think `overwriteConfig` is fitting? I can split out the logic with the defaults, like `getDefaultConfig` and `overwriteConfig`. What do you think?

Reply:
If someone has set `uid` and `gid` values, why do you want to overwrite them?

Reply:
Because we generate the uid and gid ourselves when we jail. This grants us a better security boundary between VMs, since they are all ran under a different uid and gid. We could allow for users to specify one and simply check if one had been provided, and if one has, use the given uid and gid. What do you think?

Reply:
I think you should split out the defaulting logic from this.  What does Firecracker's jailer do today?  Does it pick uid/gid for you?

Instead of overwriting values for uid/gid, can you error if those are set with a clear error message?

Reply:
No, you specify which UID and GID you want to use. However, we'd be doing the same logic, as in we'd still want to choose unique UIDs and GIDs per VM and pass that into Firecracker's jailer. This is more of an abstraction above the jailer.

> Instead of overwriting values for uid/gid, can you error if those are set with a clear error message?

Yea! I really like that